### ISSUE TITLE
Super big l2 norm of dvf gradient

### ISSUE THREAD

**Issue by XXZhou25** [voxelmorph/voxelmorph#302]
**Task** (what are you trying to do/register?)

lung CTs

**What have you tried**

Voxelmorph with MSE/NCC as loss and l2 norm of dvf gradient as regularization. 

**Details of experiments**
I did data standardization(zscore normalization with mean and std) before training. 

MSE + l2 norm of dvf gradient worked fine, as my expectation.  
But NCC + l2 norm of dvf gradient didn't work as expected. NCC was within reasonable range which is about -1~0, but l2 norm of dvf gradient was as large as hundreds. High dvf gradients mean that most of voxels' displacements are out of image, resulting in an almost blank registered moving image (interpolation will fill out-of-boundary voxels with intensity = 0).

But I used another piece of code, NCC + l2 norm of dvf gradient worked normally. 
btw, loss weight = [1, 0.5]
I was wondering if you have any insights about why? 

This is quite straight forward, computing the gradients based on Taylor expansion, which will be the difference between neighboring voxels. 
I will really appreciate if you have any suggestions!

**Comment by adalca** [comment]
@XXZhou25 Thanks for your comment on voxelmorph

Just to be clear, are you saying that the code you used, `gradientLoss` is the one that worked or the one that didn't ?

**Comment by XXZhou25** [comment]
> @XXZhou25 Thanks for your comment on voxelmorph
> 
> Just to be clear, are you saying that the code you used, `gradientLoss` is the one that worked or the one that didn't ?

Hi, sorry to probably confused you! The `gradientLoss` I used in voxelmorph, which is `voxelmorph.losses.Grad('l2')` didn't work. Actually I am not sure if it is this loss didn't work. 
More accurately, the combination: `losses = [vxm.losses.NCC().loss, vxm.losses.Grad('l2').loss]` didn't reduce the dvf loss as expected. I got out-of-image displacements. 

Btw, I am also not sure if it is my data's problem. I standardized it before training. It worked fine with my own loss code.

**Comment by adalca** [comment]
when using `[NCC, Grad]`, what was the learning rate?

**Comment by XXZhou25** [comment]
> when using `[NCC, Grad]`, what was the learning rate?
It is : 
`    lambda_param = 0.5
    loss_weights = [1, lambda_param]`

**Comment by adalca** [comment]
I mean the optimizer learning rate -- or did you use the same one that was there by default?

**Comment by XXZhou25** [comment]
> I mean the optimizer learning rate -- or did you use the same one that was there by default?

Oh I am sorry. Yeah I used Adam, with `lr = 1e-4` . 
Last time I got super large dvf, is caused by MSE + non-normalized data.

**Comment by adalca** [comment]
Hmm, this seems a bit odd, Are you sure the images are properly normalized?

Could you share a couple of screenshots (with a colorbar) for an example pair of images?

**Comment by XXZhou25** [comment]
Sure! I also tested locally, I think it is normalized correctly, Since `transformer_loss` is `<1`. Otherwise it will be large. 

First several epochs' training loss and validation loss is as below, hope it could provide more info! In epoch 0, I think `gradientLoss` was small in first several steps, but started diverge after that. 

```
epoch,disp_loss,loss,transformer_loss,val_disp_loss,val_loss,val_transformer_loss
0,90.44465,45.07303002964705,-0.14929865,2458.1616,1228.9915679931642,-0.08927019
1,3588.7134,1794.2585998535155,-0.098256014,1540.3811,770.1015716552735,-0.08902396
2,1787.0531,893.430786743164,-0.095918685,1322.4227,661.1237426757813,-0.087595165
3,1920.5934,960.1979022216797,-0.09900899,6328.1743,3163.99052734375,-0.09618848
4,1945.8949,972.8509426879883,-0.09656748,1022.1287,510.974462890625,-0.08989219
5,1199.0082,599.4064291381836,-0.09773367,772.784,386.30144653320315,-0.09054222
6,1398.1132,698.9540404129028,-0.102611616,925.3584,462.58624572753905,-0.09292896
```

**Comment by XXZhou25** [comment]
This pair is before standardize. I standardize them after reading original images.

**Comment by adalca** [comment]
The images seem to contains some machinery at the edges, interesting. I'm not sure what sort of effect this would have.

I'd recommend visualizing data in python -- the moving, fixed, moved, and the warp field (via warping a grid, maybe) with colorbars. That would help debug what is going on.

**Comment by XXZhou25** [comment]
> The images seem to contains some machinery at the edges, interesting. I'm not sure what sort of effect this would have.
> 
> I'd recommend visualizing data in python -- the moving, fixed, moved, and the warp field (via warping a grid, maybe) with colorbars. That would help debug what is going on.

Thank you very much for your help! May I show you 3D dvf? I may need to look into how to visualize with warping a grid.

**Comment by XXZhou25** [comment]
> The images seem to contains some machinery at the edges, interesting. I'm not sure what sort of effect this would have.
> 
> I'd recommend visualizing data in python -- the moving, fixed, moved, and the warp field (via warping a grid, maybe) with colorbars. That would help debug what is going on.

Thank you so much for your suggestions! It is quite odd to me, too. It worked fine with MSE or NCC with my own code. I will try to dig more! Thanks for your time!

**Comment by adalca** [comment]
I would strongly recommend visualizing middle slices them in python after they come from the generator, this would enable you to see what the network actually sees and avoid any bugs.

Have you tried training the NCC model a couple of times? you might've just been unlucky with some bad gradients at the start. Maybe try a slightly smaller learning rate as well.

**Comment by XXZhou25** [comment]
> I would strongly recommend visualizing middle slices them in python after they come from the generator, this would enable you to see what the network actually sees and avoid any bugs.
> 
> Have you tried training the NCC model a couple of times? you might've just been unlucky with some bad gradients at the start. Maybe try a slightly smaller learning rate as well.

I see. These are good points! Yeah probably data source or bad initialization problem. Will look into it! Thank you very much!

**Comment by XXZhou25** [comment]
> I would strongly recommend visualizing middle slices them in python after they come from the generator, this would enable you to see what the network actually sees and avoid any bugs.
> 
> Have you tried training the NCC model a couple of times? you might've just been unlucky with some bad gradients at the start. Maybe try a slightly smaller learning rate as well.

Hi @adalca , I just looked into everything about data, generator output, my codes...I was desperate and I just decided to run it again...Magically `gradientLoss` got back to normal right now....still not sure if it is bad initialization, since I thought probably bad, but hard not be as large as thousands. But anyways, good news is that it worked properly now! Thank you very much for your suggestions!

**Comment by adalca** [comment]
@XXZhou25 great!

**Comment by XXZhou25** [comment]
> @XXZhou25 great!

Hi @adalca , sorry I just came back. Actually dvf `gradientloss` started to go over to thousands after several epochs. Then I doubt if I should not use z-score normalization, I mean I normalized the data with its `mean` and `std`. So when I use min-max, `gradientloss` really goes back to normal. I couldn't understand why `local NCC + z-score normalization` will result in dvf loss divergence?

To be more specific, z-score I mean `(d - mean(d)) / std(d)`, min-max I mean `(d - min(d))/(max(d) - min(d))`.
Thanks!

**Comment by XXZhou25** [comment]
> @XXZhou25 great!

Hi @adalca , I am sorry I just come back again. Since I found after a while `gradientLoss` exploded to NaN. After that I also changed my normalization from Z-score standardization to Min-max, still exploded. I look into generator output, my data, data after normalization, they all looked normal. I am still confused about what caused `NCC + gradientLoss` combination diverged. I totally understand that you probably have given me biggest support from online discussion. Just want to leave some info here for possible future discussion from other people. I will update once I find out, too. Thanks!

**Comment by adalca** [comment]
@XXZhou25 which branch are you using? We updated the NCC loss in the `dev` branch to limit any NaN issues.

**Comment by XXZhou25** [comment]
> @XXZhou25 which branch are you using? We updated the NCC loss in the `dev` branch to limit any NaN issues.

Ohhh I see. I git clone from the default branch which is `master`. 
But I think I am using your `voxelmorph` library which I installed by `pip`. I am not using your scripts for losses. I mean
```
import voxelmorph as vxm
losses = [vxm.losses.NCC().loss, vxm.losses.Grad('l2').loss]
```
Is the NCC in library updated? Thank you very much!
`voxelmorph` version is `0.1`

**Comment by XXZhou25** [comment]
I just modified my losses.py from `master` branch to be consistent with `dev` branch losses. And import this scrip's loss to retrain the same model. I will update later:)

**Comment by XXZhou25** [comment]
> @XXZhou25 which branch are you using? We updated the NCC loss in the `dev` branch to limit any NaN issues.

Hi @adalca , I have run 2 experiments with `dev` branch `NCC loss` in script. I think now NCC works properly, dvf gradients never explode again. I think you probably would want to fix the `NCC` loss issue in `voxelmorph` library. It seems `NCC` isn't updated there. 

-----some update
And I also found my `transformer_loss` and `disp_loss` are both not truly reduced. My loss weights are `1, 0.5` which I tested with my own code and they worked fine, I also tried `lr = 1e-4, 1e-5`, both didn't seem to reduce the 2 losses (more specific, `NCC(dev branch)` + `Grad(master branch)` . Not sure if it is the problem of hyper-parameter or I should update `Grad` too. I will update all losses to `dev branch` and retrain. Thanks.

**Comment by adalca** [comment]
Thanks for the update. We will definitely push the new NCC to pypi when we test everything.

What do you mean by 'truly reduced' ?

Also regarding hyperparameters, you might want to look into https://arxiv.org/abs/2101.01035

**Comment by XXZhou25** [comment]
> Thanks for the update. We will definitely push the new NCC to pypi when we test everything.
> 
> What do you mean by 'truly reduced' ?
> 
> Also regarding hyperparameters, you might want to look into https://arxiv.org/abs/2101.01035

Thank you! Now I am using all losses from `dev` branch. `MSE + Grad` seems to work fine, total loss is decreasing. But `NCC + Grad`, total loss is not decreasing. I will look into that paper and also keep parameter searching! Thanks!
![Screen Shot 2021-05-02 at 8 44 34 AM](https://user-images.githubusercontent.com/48270633/116815338-a7f4ad00-ab22-11eb-82c7-9aa54ee95c61.png)

**Comment by adalca** [comment]
How long have you run this for? In a lot of situations in deep learning, you can be in these wide saddle points where the loss doesn't change for a while, then starts to drop. This can happen because of the loss landscape and will differ with the learning rate, etc. I would recommend trying to run it for a long while and see if this helps. you can also try making the learning rate a bit higher, or initialize an NCC loss after you've trained with MSE for a bit.

**Comment by XXZhou25** [comment]
> How long have you run this for? In a lot of situations in deep learning, you can be in these wide saddle points where the loss doesn't change for a while, then starts to drop. This can happen because of the loss landscape and will differ with the learning rate, etc. I would recommend trying to run it for a long while and see if this helps. you can also try making the learning rate a bit higher, or initialize an NCC loss after you've trained with MSE for a bit.

Hi @adalca , I have a question about your NCC in `dev` implementation. 
`cross = IJ_sum - u_J * I_sum - u_I * J_sum + u_I * u_J * win_size`
     why multiply by `win_size` in last term? I mean why it is not `u_I * u_J `?
     According to NCC's denominator `(f - f_mean)(g - g_mean) = fg - f_mean * g - f * g_mean + g_mean * f_mean`. 
I believe `u_I, u_J` have been the local mean. 
Same question for 2 variance terms, which are:
```
I_var = I2_sum - 2 * u_I * I_sum + u_I * u_I * win_size
J_var = J2_sum - 2 * u_J * J_sum + u_J * u_J * win_size
```

Thank you so much!

**Comment by adalca** [comment]
I'll leave this one to @balakg  :)

**Comment by XXZhou25** [comment]
Hi, may I ask one probably super basic question? 
I think global NCC's range should be `[-1, 1]`, local NCC should be the same, since I think in last step, local NCC would take average of all windows's NCC. 
But actually in my experiment, transformer_loss (local NCC, window_size = 9) could reach to `-57.63151`, any number smaller than `-1`.

**Comment by XXZhou25** [comment]
Sorry, I was wondering if anyone could clarify a little bit? I'll really appreciate.

**Comment by adalca** [comment]
@XXZhou25 are you using the new NCC in `dev`? I can't say that we've ever seen the NCC loss be higher than -1 honestly, but perhaps there are some corner cases or a small bug? What images were you using?

**Comment by XXZhou25** [comment]
> @XXZhou25 are you using the new NCC in `dev`? I can't say that we've ever seen the NCC loss be higher than -1 honestly, but perhaps there are some corner cases or a small bug? What images were you using?

@adalca Yes, I am using new NCC in `dev`. Here is my copy of your code, I didn't modify anything. I used Dirlab and POPI dataset, which are patient lung 3D CT. 

```
class NCC:
    """
    Local (over window) normalized cross correlation loss.
    """

    def __init__(self, win=None, eps=1e-5):
        self.win = win
        self.eps = eps

    def ncc(self, Ii, Ji):
        # get dimension of volume
        # assumes Ii, Ji are sized [batch_size, *vol_shape, nb_feats]
        ndims = len(Ii.get_shape().as_list()) - 2
        assert ndims in [1, 2, 3], "volumes should be 1 to 3 dimensions. found: %d" % ndims

        # set window size
        if self.win is None:
            self.win = [9] * ndims

        # get convolution function
        conv_fn = getattr(tf.nn, 'conv%dd' % ndims)

        # compute CC squares
        I2 = Ii * Ii
        J2 = Ji * Ji
        IJ = Ii * Ji

        # compute filters
        in_ch = Ji.get_shape().as_list()[-1]
        sum_filt = tf.ones([*self.win, in_ch, 1])
        strides = 1
        if ndims > 1:
            strides = [1] * (ndims + 2)

        # compute local sums via convolution
        padding = 'SAME'
        I_sum = conv_fn(Ii, sum_filt, strides, padding)
        J_sum = conv_fn(Ji, sum_filt, strides, padding)
        I2_sum = conv_fn(I2, sum_filt, strides, padding)
        J2_sum = conv_fn(J2, sum_filt, strides, padding)
        IJ_sum = conv_fn(IJ, sum_filt, strides, padding)

        # compute cross correlation
        win_size = np.prod(self.win) * in_ch
        u_I = I_sum / win_size
        u_J = J_sum / win_size
        # u_I, u_J has normalized the cross correlations

        # TODO: simplify this
        # cross = fg - f_mean * g - g_mean * f + f_mean * g_mean
        cross = IJ_sum - u_J * I_sum - u_I * J_sum + u_I * u_J * win_size
        cross = tf.maximum(cross, self.eps)
        I_var = I2_sum - 2 * u_I * I_sum + u_I * u_I * win_size
        I_var = tf.maximum(I_var, self.eps)
        J_var = J2_sum - 2 * u_J * J_sum + u_J * u_J * win_size
        J_var = tf.maximum(J_var, self.eps)

        # cc = (cross * cross) / (I_var * J_var)
        cc = (cross / I_var) * (cross / J_var)

        # return mean cc for each entry in batch
        return tf.reduce_mean(K.batch_flatten(cc), axis=-1)

    def loss(self, y_true, y_pred):
        return - self.ncc(y_true, y_pred)
```

**Comment by XXZhou25** [comment]
@adalca For more information, here is one validation result:
`epoch,disp_loss,loss,transformer_loss,val_disp_loss,val_loss,val_transformer_loss
0,0.15083538,-49.949016827661545,-50.024437,0.45199376,-73.46460145264864,-73.6906
1,0.3055039,-22.163731549941005,-22.316484,0.42120677,-0.48982425332069396,-0.7004276
2,2.3713417,-143.19856023862957,-144.38422,0.6123885,-18.74362783059478,-19.049824
3,0.83364713,-73.94608908504247,-74.36291,7.1147127,-3.578283405303955,-7.13564
4,3.6995177,-70.77165074169636,-72.621414,1.416628,-68.6590909332037,-69.36741
5,2.5511255,-54.558479174822565,-55.834057,1.6698322,-43.24957476258278,-44.08449
6,0.6474533,-93.70873383037745,-94.03247,0.7719676,-47.848022386431694,-48.234005
7,1.243567,-55.54088136941195,-56.16267,0.66736215,-6.433693189173937,-6.7673745
8,1.7654729,-55.217502191364765,-56.100254,1.093515,-44.880161659419535,-45.426918
9,1.2122403,-34.02161846697331,-34.62774,0.98959255,-85.13306598812342,-85.62785
10,1.0766712,-80.36594043545425,-80.904274,0.65233266,-51.31651806533337,-51.64269
11,0.6765165,-45.36293393306434,-45.70119,0.74835265,-1.131783763319254,-1.5059601
12,2.4912322,-101.19795673549176,-102.44357,0.6611043,-34.762744311243296,-35.0933
13,1.0940706,-107.655750688985,-108.202774,1.6965386,-73.43013369888067,-74.2784
14,1.6649861,-88.92588941134512,-89.75838,2.7781231,-56.242446404695514,-57.63151
15,1.3063091,-91.56569169268012,-92.21884,1.1813835,-3.3971395045518875,-3.987831
16,1.1905783,-108.69550204455852,-109.29081,2.7854636,-3.7221172496676447,-5.114849
17,1.7749962,-53.27971044436097,-54.167206,0.6964264,-1.7472069911658763,-2.0954201
18,2.1475167,-66.20176459684969,-67.275536,0.660285,-35.78572437986732,-36.115868
19,1.2830027,-148.1695452737808,-148.81107,0.9387504,-64.27786657586694,-64.747246
20,0.7700645,-40.42144216105342,-40.80647,0.62782675,-8.620680951327085,-8.934594
`

**Comment by XXZhou25** [comment]
For this experiment, I used dirlab + POPI datasets, they are both 3D CTs, and include masks to set background voxels to be 0 and only extract human body part, which means: 
`
image = (image - mean ) / std,
humanbodypart = image * mask
`
However, with same dataset and same everything except learning rate, when I decrease initial learning rate , from `1e-4` to `1e-5`, `transformer_loss` got back to normal, in a reasonable range. 

But when I am using much more data, which include 8 datasets but didn't use any masks(i.e. model input is original normalized images), the `transform_loss` diverge to as large as 130+, as well. 
Therefore, I am not sure if it is caused by data? 
I thought NCC is based on local correlation, will diverse dataset cause NCC to diverge?

**Comment by XXZhou25** [comment]
> @XXZhou25 are you using the new NCC in `dev`? I can't say that we've ever seen the NCC loss be higher than -1 honestly, but perhaps there are some corner cases or a small bug? What images were you using?

Yes I include masks to extract human body parts for each image, in this way, when register 2 images, there will be some 'edges', not sure if they are corners. Will NCC be affected by this? Thank you so much!

**Comment by adalca** [comment]
Sorry for not responding earlier, but I don't think we've seen this before.

If you just load up two images and evaluate our NCC on those images, is it within 0 and 1?
